GitHub integration

sgfost opened the pull request:
adds a feature to the CML which allows for:
- **connecting** a public GitHub repository with a codebase in the CML and;
- **pushing** pushing a git repository constructed from public CML releases to the connected GitHub remote
- **importing** individual releases made on a GitHub repository into the CML as new releases (of a special "imported" type) which can be published/peer reviewed

Setup/management is done by the submitter on a configuration page for each codebase, accessible from the sidebar of a codebase page or the feature overview page (`/github/`). From there, one is walked through the process including connecting a github account with the existing oauth flow and installing our github app on the same account, then given the option of:
- connecting with a **new repository**: allows connecting to a blank github repository, then generates a git repo for a codebase which can be pushed (codebase updates are reflected in the git repo and the push can be repeated). Importing GitHub releases back is also allowed
- connecting with an **existing repository**: allows connecting with an existing GitHub repo and importing GitHub releases as codebase releases

### configuration

-> https://github.com/comses/infra/wiki/GitHub-Sync

### details

#### new models

- `GithubIntegrationAppInstallation`: tracks the github app installation status for a user
- `CodebaseGitRemote`: represents a remote repository that should be pushed to and/or checked for new releases
- `GitRefSyncState`: represents the state of a git ref (commit/tag/branch) that was generated for a codebase and the state of the push to GitHub
- `ImportedReleaseSyncState`: represents the state of a release imported from GitHub, the existence of this means a release was imported from github, and should be treated differently in some cases
- `GitHubIntegrationConfiguration`: (admin dashboard-editable) feature settings including overview page content

#### github APIs

- `GitHubApi`: main wrapper for github operations (auth, pushing, creating releases, etc.)
- `GitHubReleaseImporter`: handles importing releases from github by downloading the archive, extracting metadata, and creating a new CodebaseRelease

#### filesystem

- `CodebaseGitRepositoryApi`: manages local git repository operations like building and updating a git repo from a codebase's (public) releases
- `FileCategories`/`CategoryManifestManager`: file categorization changed in order to handle imported releases which are not organized into category directories, information is instead kept in a manifest in the db and updated by the submitter
- `ImportedCodebaseReleaseFsApi`: sibling filesystem API of the regular `CodebaseReleaseFsApi` with some special handling and additional functionality
- adds generated `LICENSE` and `CITATION.cff` files to new release packages

#### webhooks

`library.views.github_sync_webhook`: the github app sends webhook events to the server and we respond to installation events to track user installations

sgfost committed:
feat: add model for tracking git mirror of a codebase

sgfost committed:
feat: add codebase git repository fs api

this API is responsible for managing a local git repository mirror for a
comses codebase. PUBLIC release archives are commits/tags in the
history. Release branches are created for each release and only added to
if there is an update to metadata

`build()` and `append_releases()` are the two main API methods which
construct (or rebuild) a git repo and add new releases to the repo,
respectively

`update_release_branch()` will add a new commit containing changes to a
release branch (and update main if they point to the same thing). This
will mainly be used for updating metadata

sgfost committed:
feat: add github integration api and mirroring tasks

the GithubApi provides access to auth and repository actions

adds 3 huey (async) tasks for creating a mirror, updating a mirror, and
updating metadata for a single release of a mirror

sgfost committed:
feat: github integration ui and mirroring feature

* /github page to describe the integration features
* sidebar element on release detail page will show information about
  integration status for that codebase, and allow users with edit
  permissions to create a new mirror

sgfost committed:
feat: rework git integration models and github/fs apis

* use installation access tokens for user repos instead of user access
  tokens. this is a more secure workflow
* add GithubIntegrationAppInstallation model for recording app
  installations (this will need to be created/updated using webhooks)
* CodebaseGitMirror/"mirror" now refers to the local git repository
* ^ can have multiple CodebaseGitRemote's which keep track of all the
  information needed to push to/archive from remote repositories

TODO: re-implement views

sgfost committed:
feat: views/routing for github integration config

sgfost committed:
feat: add github sync configuration page

this replaces the simple modal form to give better control over the feature

sgfost committed:
feat: add create/update functionality to the gh sync config

sgfost committed:
feat: add guidance for connecting acct in gh sync config

sgfost committed:
feat: add new feature overview/explainer page for gh sync

sgfost committed:
feat: add external/archived release type

The distinguishing feature is whether the release has a non-null
external_release_package

This will be used to 'archive' or pull in releases made on github for
synced repos

currently, the release assets/package is not stored on the filesystem,
instead relying on an external download url, and being only concerned
with metadata

sgfost committed:
feat(WIP): user-owned repos webhook + workflows

App installation tokens did not give access to get/create repos on a
user's account. Still trying to avoid storing user access tokens
(https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-user-access-token-for-a-github-app)
so as a workaround, we will direct the submitter to create a new bare
repo before continuing

* add handler for webhook events for the github sync app
* add form/wizard for linking a pre-existing github repo (archive only)

sgfost committed:
feat: user-owned repos are working

allows setting up a push/archive sync that will automatically have the
generated git repo pushed to by providing an empty repo, as well as
setting up an archive only sync by providing any github repo

in both cases, the submitter needs to:
- link their github account with the regular oauth flow
  (so we have a way to match users with a github account)
- install the provided 'GitHub Sync' app on their github account with
  access to any repository that will be synced

sgfost committed:
feat: remove github sync org option, refine UI and overview/help page

sgfost committed:
fix: rename github -> CML process 'archiving' -> 'importing'

and squash migrations

"import(ing)" is the wording I keep finding describes the process the
best

other potential names and their issues:
* publish - same name as the direct publishing, releases need to be
  manually published after they are imported
* pull - git command that is not used
* fetch - git command that is not used
* ingest - ok and similar to import but not quite as clear

sgfost committed:
fix: improve syncing workflow and transparency

* re-order and clarify the steps to set up a sync (app installation
  takes place after creating a repo so that permissions can be
  restricted)
* fix push log to actual show useful information
* when toggling push back on, do a build/update + push on the spot
* better error/success messages

sgfost committed:
feat(WIP): github release importing

sgfost committed:
feat(WIP): add release FS API for imported releases

refactors the `CodebaseReleaseFsApi` to inherit from an abstract
`BaseCodebaseReleaseFsApi`, along with the new
`ImportedCodebaseReleaseFsApi`

the main function of the imported release fs api is to import releases
from a remote source by downloading an archive to originals/, extracting
to sip/data/ and then using the inherited functionality to build
archives for review/publishing. It does not implement methods for
dealing with files directly outside of this importing workflow like
add(), delete(), rebuild(), etc.

the imported release fs api manages a manifest for keeping track of file
categories. Using this for all releases is potentially something to do
in the future but is a rather complicated refactor that had to be dialed
back here

sgfost committed:
feat: github release importing

* webhook handler watches for github release events and creates a db
  record and delegates to the FS api to download and extract a release
  archive into the library FS

TODO:
- tailor the release editor UI for imported releases, including a
  way to categorize files
- extract initial metadata from the github release archive

sgfost committed:
feat: imported-specific release editor UI

including file categorization with an improved file tree

sgfost committed:
feat: extract metadata from imported releases to populate CML release

currently pulls out: license, release notes, languages, platforms, os
from: github repo/release data, any found codemeta file, any found
      CITATION.cff file

sgfost committed:
fix: add configurable github app webhook secret

and fix some typescript complaints

sgfost committed:
feat: add gh sync config settings model, clarify overview

and disallow syncing with private repos for now. adds too much
complexity to manage if we allow this

FAQs are now configurable through the settings model for the feature
along with a beta toggle and whether to enable setting up new syncs

a base set of FAQs is created in the migration

sgfost committed:
feat: improve gh sync configuration UI

* added back CodebaseReleaseFsApi get_sip_list_url and
  get_originals_list_url helpers. I'm not entire sure why these got removed

sgfost committed:
refactor: github install status method for any user

in addition to the submitter of a specific codebase, this is so we can
display this info/wizard thing outside the context of any given codebase

* also give some context to the old external repository url field (just
  a reference)

sgfost committed:
feat: add archive preview in release editor for published releases

this solves the issue where imported + published releases were trying to
show the progress without access to the original files by adding a
read-only archive preview for both types of releases

sgfost committed:
fix: resolve tags when importing releases

adding tags by string name is unsafe when there are (case-insensitive)
duplicate tags in the db

sgfost committed:
feat: improve github sync overview page

by splitting up into tabs and giving direct access to the feature by
providing a list of the user's submitted models to configure a sync for

* fixed an issue where importing couldn't be set up for unpublished
  codebases

sgfost committed:
feat: slim down model library preamble thing

+ add link to github sync main page

sgfost committed:
feat: github sync config app link and soft feature flag

sgfost committed:
fix: clarify and correct coerce_codemeta/cff

coerce_codemeta was incorrectly raising an exception if codemeta=None
(and codebase=None, release=None) was given when it should simply fall
through

sgfost committed:
fix: imported release peer review

- fixed bug preventing creating new drafts when the last release is an
  imported release
- disable requesting review (with note) on imported+published releases
- add instructions in email for making revisions on an imported review

sgfost committed:
test: add github sync importer/webhook/etc tests

* fixed issue with re-imported releases not updating the release package
  info (downloadurl, tag name, etc.)
* refactor metadata extractor for easier priority changes

sgfost committed:
fix: dont try to create review draft copy of unpublished releases

in other words, only create draft copies when the release is published,
unpublished (how imported releases come in) should work the same as
drafts

also added some extra safety checks for requesting review on a release
is that already under review or completed review

sgfost committed:
fix: conditionally send diff versions of revisions requested email

sgfost committed:
feat: add some indicators for imported releases

on the model detail page as well as in review stuff

sgfost committed:
feat: switch to a manual release syncing workflow

rather than opaque automatic syncing when a release is published on
either end of the cml<->github integration, 'imports' and 'pushes' are
manually triggered in a management view

configuration of a github connection is now in a much more streamlined
wizard-style set of steps with better feedback

sgfost committed:
refactor: simplify outbound git ref tracking

a simpler 1:1 representation of the git refs (commit/branch/tag) and
associated push state replaces the previous N:1 which would create
multiple records for a single release to represent a connection with a
remote

since remote repo swapping will be rare, this greatly cuts down on
complexity and fixes several bugs at the cost of losing metadata when
swapping remotes

also added support for tracking the main branch with the same
GitRefSyncState model

sgfost committed:
fix: prevent recursive update loops when building git repos

also, prevent multiple concurrent push jobs and fix up some wording

sgfost committed:
feat: better github integration release management

now using a table that visually pairs releases on either side

* also added support for going straight to the github integration setup
  from codebase creation form, for people who would like to import their
  work from there

sgfost committed:
fix: migration, mark github integration info page as outdated

[no ci]

sgfost committed:
feat: another gh integration overview page rework

pending initial seed migration for some of the db-driven content

sgfost committed:
content: initial github overview page with placeholder videos

sgfost committed:
fix: imported release metadata extraction + ui fixes

- incorporate upstream fix in codemeticulous, now install from specific
  commit on SciCodes repo, still pending pypi release
- on a new release, save before generating codemeta since the full
  generation would always fail without a pk
- fix a couple type mismatches in the fs module and settings

sgfost committed:
feat: more info for imported releases under review

also fix some potential type errors in importing

sgfost committed:
fix: additional note on "pushing" to github

sgfost committed:
fix: improve messaging and feedback for github connections

- give hint if no github releases are found for connected repo
- add repo connection success message with info about returning to the
  management page
- fix issue where deactivated remotes couldn't be re-activated for the
  same codebase
- link to/explain how to submit new models straight from github
- spell out more requirements for making new repos and disable
  continuing until "create a new blank repo" is clicked on

sgfost committed:
fix: include git repositories in borg repo

sgfost committed:
test: update github integration tests

based on the last large refactor that involved changing up the data
model

sgfost committed:
fix: validate integration app access before connecting to repos

sgfost committed:
fix: fail main branch push jobs when they fail

sgfost committed:
feat: gh integration feature flag and repo disconnection button

includes a global feature flag and allowed users list (ignored when
globally enabled, superusers are implicitly allowed)

replaces swapping repository connection workflow with a disable button,
which allows for the same thing but in discrete steps

+ some guidance on edge cases in FAQ

sgfost committed:
fix: github release importer uses the correct programming langs model

+ tidying migrations

sgfost committed:
fix: some minor issues/bugs

sgfost committed:
fix: github integration app id should be a str

pyjwt expects it as such

alee committed:
ui: consider text wrap badge instead of truncate

alee commented on deploy/conf/.env.template:
```diff
@@ -45,6 +45,12 @@ DATACITE_DRY_RUN="true" # allowed values: "true" or "false"
+# github integration app
+GITHUB_INTEGRATION_APP_ID=
+GITHUB_INTEGRATION_APP_NAME=
+GITHUB_INTEGRATION_APP_INSTALLATION_ID=
```
should document INSTALLATION_ID on the infra wiki: purpose, how / when is it set in dev / staging / prod, etc

alee commented on django/core/tests/base.py:
```diff
@@ -187,5 +187,13 @@ def initialize_test_shared_folders():
+        if fs.is_dir():
+            shutil.rmtree(os.path.join(dir, fs.name), ignore_errors=True)
+        elif fs.is_file():
+            os.remove(os.path.join(dir, fs.name))
```
should standardize new code to favor pathlib

alee committed:
refactor: prefer pathlib to os for fs operations

alee committed:
refactor: favor pathlib.Path over os.path

alee committed:
refactor: minor test refinement

add targeted test selection through make and deploy test script
`make test TEST_ARGS="library.tests.test_views"

return 400 for malformed GitHub installation payloads, add guards for
JSON parsing, and add tests

add version collision test to make sure duplicate version imports are
rejected and assert no duplicate release / package side effects on
collision

minor metadata test refinement

alee committed:
refactor: minor refinement

- centralize remote token access and fail fast
- add Codebase.code_repository to check for git remotes and then
  fallback to repository_url

alee committed:
refactor: extract structural repo check

alee committed:
refactor: minor test hygiene

alee committed:
refactor: add logging

alee commented on django/library/models.py:
```diff
@@ -3347,3 +3883,166 @@ def __str__(self):
                    """
+
+
+class GitHubIntegrationFaqEntry(index.Indexed, Orderable):
```
Minor nitpick that is a non-blocker, we should be consistent about naming (e.g., `GithubIntegrationAppInstallation`). I think I tend to favor `Github` for Python symbols to NotCapitalizeAllTheThings and GitHub in docs etc. but since there are many more of these it's probably easier to just rename `GithubIntegrationAppInstallation` and then redo the schema migrations.

alee commented on frontend/src/components/githubIntegration/AccordionStep.vue:
still not a fan of camel casing in the directory (`releaseEditor` suffers from the same issue), would prefer `release-editor` and `github-integration` on the filesystem

alee commented on frontend/src/components/githubIntegration/ReleaseManagementSection.vue:
```diff
@@ -0,0 +1,406 @@
+  return Number.isNaN(t) ? 0 : t;
+}
+
+function sortEntries(matrix: ReleaseAlignment) {
```
might be worth bringing in the npm `semver` or `semver-ts` library for this instead if we want to deal with semver corner cases but this is probably fine for initial release

alee committed:
docs: refine upload instructions

change docker compose flags to --progress=plain and build --parallel,
remove -q to see debug information

alee committed:
fix(WIP): add logging to github_sync_webhook

alee committed:
chore: remove logs and backups from docker context
